[Wang Haoyang] iP

haoyangw opened the pull request.

haoyangw committed:
Don't include .class files generated during testing in git repo

haoyangw committed:
Level 0: Duke shall greet the user and exit, for now

haoyangw committed:
Implement increment Level-0: Read commands and reply

haoyangw committed:
Add some personality to (who is now) R.I.C.A

haoyangw committed:
Implement Level-2: Add and List functionality for R.I.C.A

haoyangw committed:
Implement a basic Task class that stores task status and description

haoyangw committed:
Refactor Rica to use Task objects(basic) instead of Strings

haoyangw committed:
Implement logic for marking tasks as done through Rica

haoyangw committed:
Implement logic for marking tasks as not done while fixing a bug in parsing command as mark/unmark

haoyangw committed:
Augment list command: Tell the user when no tasks have been added yet, otherwise list tasks

haoyangw committed:
Improve the reply when marking/unmarking tasks as done

haoyangw committed:
Move reply for mark/unmark command to the corresponding helper function

haoyangw committed:
Bounds checking: Check that task exists before trying to mark/unmark to prevent crash

haoyangw committed:
Sanity check: Don't mark/unmark task when it's already marked/unmarked, also inform the user about it

haoyangw committed:
Add proper padding for mark/unmark reply that matches that of other commands for tasks

haoyangw committed:
Another sanity check: If user spams whitespace/enter key don't end up adding a task with whitespace name/no name

haoyangw committed:
Add javadoc comments for all non-trivial methods in our classes

haoyangw committed:
Rica.class: Remove unused overloaded addTask() method

ChongQiRong commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    public static void main(String[] args) {
+        // Define horizontal line in a variable since it's printed multiple times
```
Can remove redundant comment

ChongQiRong commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,36 @@
+public class Task {
+    protected final String DESCRIPTION;
```
Variable naming should be camelcase

ChongQiRong commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+
+    public static void main(String[] args) {
+        // Define horizontal line in a variable since it's printed multiple times
+
```
Can remove redundant line

ChongQiRong commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+                }
+            } else {
+                switch (command) {
+                    case Rica.LIST_TRIGGER:
```
case should be inline with switch

ChongQiRong reviewed (commented):
Good refactoring of code with minor coding standard errors

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+     * @param indexOfTask Index of desired task in the task list
+     * @return Task object representing the desired task being marked as not done
+     */
+    private static Task unmarkDone(int indexOfTask) {
```
unmarkDone method is relatively long and can be refactored into smaller method

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+     * @param indexOfTask Index of given task in the task list
+     * @return Task object representing the desired task being marked as done
+     */
+    private static Task markDone(int indexOfTask) {
```
markDone method is relatively long and can be refactored into smaller method

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+    private static final String UNMARK_TRIGGER = "unmark";
+    private static final String BYE_TRIGGER = "bye";
+    private static final String ADD_PHRASE = " New task I'll remember: ";
+    private static final String BYE_PHRASE = " Leaving so soon? Come back anytime, I'll be happy to help!";
```
Good to initialize string constants as private static final

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+            printlnWithIndent(" Hope I'm not amnesiac, but I don't remember any tasks?");
+        } else {
+            ArrayList<Task> tasks = Rica.getPastTexts();
+            printlnWithIndent(" I think you have these tasks:");
```
Hardcoded strings can be refactored into constants at the top of code for maintainability

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+import java.util.ArrayList;
+import java.util.Scanner;
+
+public class Rica {
```
Rica class is doing multiple things such as adding marking and printing tasks and separating into different classes can be used for better clarity

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+        return selectedTask;
+    }
+
+    /**
```
Comments can be evenly placed into the individual lines of code for better clarity on what each line of code does

DaDevChia commented on src/main/java/Rica.java:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    private static boolean hasAnyTasks() {
+        return !Rica.getPastTexts().isEmpty();
```
Double negative in logic leads to less clarity

DaDevChia reviewed (commented):
For Task 1 of PR

haoyangw committed:
Task.class: Fix non-final class variable name being written in caps instead of camelCase

haoyangw committed:
Rica.class: Fix indentatino for switch block

haoyangw committed:
Task.class: Follow coding style standards and use just one setDone() setter

haoyangw committed:
Add a new Todo class and move todo logic from Task to Todo

haoyangw committed:
Rica.class: Refactor to use Todo class instead of Task

haoyangw committed:
Rica.class: Add a new command for adding todos

Also, rename our variables and methods now that we are using Todos instead of Tasks directly

haoyangw committed:
Todo.class: Add additional [T] to its toString() output

haoyangw committed:
Task.class: Make getDescription() publicly accessible

Not sensitive information anyway, in fact it is quite meaningful for other classes

haoyangw committed:
Move logic for creating new Todo to a factory method in Todo.class

haoyangw committed:
New TaskManager.class: Manages tasks(add, mark, unmark)

haoyangw committed:
Rica.class: Refactor to use the new TaskManager.class

Call TaskManager's methods to perform task management so Rica.class is simpler.

haoyangw committed:
Rica.class: Fix warnings due to using Integer.valueOf() wrongly

haoyangw committed:
New Deadline.class: Todos that must be done before a date/time

haoyangw committed:
Update Rica.class and TaskManager.class: Support adding deadline tasks

haoyangw committed:
New Event.class: Support adding events with start/end date/time

haoyangw committed:
Update Rica.class and TaskManager.class: Support event commands by using Event.class methods

haoyangw committed:
Fix additional whitespace messing up formatting of task management replies

haoyangw committed:
TaskManager.class: Fix refactoring misses and grammar issue

haoyangw committed:
TaskManager.class: Improve if conditions' readability

Store boolean conditions in variables with descriptive names so if conditions are easily understandable and shorter

haoyangw committed:
Code quality improvements

Set some rules for arranging variables/methods in all classes from now on

1. Order of arranging: Variables first, followed by constructors, and then methods
2. Within each of the above categories, arrange by their access modifier: private->default->protected->public
3. Within each of the access modifier categories, arrange names alphabetically

haoyangw committed:
Code quality improvements: Reduce method length

Abstract logic into separate methods more to reduce the length of each individual method

Makes code more readable

haoyangw committed:
Add our own RuntimeException class

Have RicaException to represent all of Rica's internal exceptions and RicaTaskException(subclass of RicaException)
  to represent Rica's task-related exceptions

haoyangw committed:
Use RicaException and RicaTaskException to handle errors gracefully

Now Rica won't crash or return null references anymore when given invalid inputs

haoyangw committed:
Merge branch 'branch-Level-5'

Implement increment Level-5

haoyangw committed:
Create two preliminary packages: exception and task

exception: For RicaException and its subclasses to categorise exception conditions in Rica

task: For Task and its subclasses to categorise Rica's various tasks

haoyangw committed:
Merge branch 'branch-A-Packages'

Implements the increment 'A-Packages'

haoyangw committed:
TaskManager.class: Implement removing task logic

Actually we already have an internal helper method for removing a task by index, so just implement command validation and then
  let the helper method rmTask() do its job

haoyangw committed:
Rica.class: Support a new command, 'delete'

Let TaskManager.class handle the task deletion logic using its rmTask() method

haoyangw committed:
TaskManager.class: Arrange rmTask() by access modifier and alphabet

Accidentally put the publicly-accessible rmTask() below the private rmTask(), oops. Broke the method arrangement rules I set earlier

haoyangw committed:
Task.class: Implement getType() as well to facilitate serializing

haoyangw committed:
Todo.class: Use @Override tag on getType()

Now that getType() is also implemented by Task.class, make this clear by marking getType() with the @Override tag

haoyangw committed:
TaskManager.class: Refector all static methods and variables

Fix a bad abstraction done earlier where static methods from Rica were copied to a new class(TaskManager) without removing the
  unnecessary 'static' modifier.
After all, all of TaskManager's methods and variables are now called from an instance of TaskManager created by Rica, so marking them
  'static' is plain wrong

haoyangw committed:
Add new Serializable interface

Our way of marking classes as Rica-serializable. Will have to implement serializeObject() method that returns a String representation
  of the object that we can save

haoyangw committed:
Add new RicaException: RicaSerializationException

Categorises an internal Exception as serialization-related

haoyangw committed:
Add another RicaException: RicaStorageException

Categorises an internal Exception as data storage related. Lays the ground work for saving our task list later

haoyangw committed:
Add new StorageManager.class to handle storing Rica's data

As the name implies. Abstract out logic for reading/writing data to
  persistent storage into this class.

Default data file is {project directory}/data/tasks.txt, though a
  different path may be specified with the overloaded constructor.

If any directories along the data file path don't exist, it'll be
  created automatically(with warning, but no crash).

StorageManager.class throws RicaStorageException when it encounters
  exceptions reading/writing to persistent storage.

haoyangw committed:
Todo.class: Implement our Serializable interface

Allow todos to be serialized into Strings using Todo.serializeObject(),
  and add a factory method that recreates a Todo instance from a
  valid serialized String(Todo.deserializeObject())

haoyangw committed:
Todo.class: Properly encapsulate isDone instance variable

Make isDone a private variable so childclasses have to use
  corresponding getter/setter methods to access/update it
  (the proper OOP way)

Also, to support instantiating childclasses from a serialized
  String, childclasses need to set the value of isDone from
  their constructors. Expose the private Todo() constructor
  that sets isDone value to childclasses as well to support
  this behavior.

haoyangw committed:
Todo.setDone(): Use the correct scope for getDescription()

getDescription() is defined in the superclass Task.class,
  so make this obvious by calling it via the super reference
  instead of the this reference

haoyangw committed:
Task.class: Implement the Serializable interface

Mark Task.class as Rica-serializable by implementing the
  serializeObject() method from Serializable.

Also, provide a factory method deserializeObject() that
  instantiates a valid Task instance from a serialized
  String representation

haoyangw committed:
Event.class: Implement the Serializable interface

Add a private constructor for instantiating all instance
  variables from given values to support deserializing

Like for its superclasses, implement serializeObject() and
  provide a factory method for instantiating an instance of
  Event from a serialized String representation(that uses
  the new private constructor).

haoyangw committed:
Deadline.class: Implement Serializable interface

Like Event.class, add a private constructor for
  instantiating all instance variables with given
  values to support deserializing.

Then we implement serializeObject() and provide a
  factory method for instantiating an instance of
  Deadline from a serialized String representation.

haoyangw committed:
Task.class: Augment its deserializeObject() method

As the superclass, let Task.deserializeObject() handle(and
  delegate) the deserialization of Task and its subclasses
  from local storage.

Throw RicaStorageException if an unrecognised Task-related
  object data is given.

haoyangw committed:
Fix bug in marking deadlines/events as done

Deadlines/Events would get turned into plain Todos when
  marked done.

Fix by providing overridden methods in Deadline/Events
  that return the correct Task subclass when marked done.

haoyangw committed:
TaskManager.class: Load saved tasks on initialization

Create an instance of StorageManager in TaskManager's constructor
  and let StorageManager load saved tasks into an ArrayList.

If an exception occurs, just initialise an empty ArrayList and
  pretend no tasks were previously saved.

haoyangw committed:
TaskManager.class: Save tasks on any task list change

Every time a Task is added, marked done, or marked as not done,
  save the task list through StorageManager.

To be specific, StorageManager overwrite's the content of our
  data file with the new task list.

haoyangw committed:
git: Ignore Rica's default local storage folder

Don't add any files in data/ into git

haoyangw committed:
git: Also ignore checkstyle config folder

haoyangw committed:
Merge branch 'branch-Level-6'

Implements increment Level-6

haoyangw committed:
Merge branch 'branch-Level-7'

Implements increment Level-7

haoyangw committed:
TaskManager.class: Save task list on task deletion

Didn't manage to implement this in Level-7 since
  Level-6 and Level-7 are in different branches.

This should make the task list persistence robust.
  Task list is saved in all known cases.

haoyangw committed:
git: Don't add jar metadata(META-INF) to repo

qianz-z commented on src/main/java/io/github/haoyangw/rica/Rica.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Scanner;
+
+public class Rica {
+    private static final String BYE_PHRASE = " Leaving so soon? Come back anytime, I'll be happy to help!";
```
Good use of constants in your code!

qianz-z commented on src/main/java/io/github/haoyangw/rica/Rica.java:
```diff
@@ -0,0 +1,94 @@
+            case Rica.DEADLINE_TRIGGER:
+            case Rica.EVENT_TRIGGER:
+            case Rica.TODO_TRIGGER:
+                try {
```
Try to avoid deep nesting here :>

qianz-z commented on src/main/java/io/github/haoyangw/rica/task/Deadline.java:
```diff
@@ -0,0 +1,112 @@
+                }
+                break;
+            }
+            if (i != 1) {
```
Try to avoid magic literals. There has been a few more instances in the rest of your code. You can use a named constant makes the code easier to understand because the name tells us more about the meaning of the number.

qianz-z commented on src/main/java/io/github/haoyangw/rica/task/Deadline.java:
```diff
@@ -0,0 +1,112 @@
+    public static Deadline create(String command) throws RicaTaskException {
+        String[] parameters = command.split(" ");
+        // Wrong command for adding a deadline
+        if (!parameters[0].equals(Deadline.COMMAND)) {
+            throw new RicaTaskException(Deadline.WRONG_CREATE_CMD_ERROR);
+        }
```
Good and prominent happy path! Good job!

qianz-z commented on src/main/java/io/github/haoyangw/rica/task/TaskManager.java:
```diff
@@ -0,0 +1,240 @@
+    private Task getTask(int indexOfTask) {
+        boolean isNegativeIndex = indexOfTask < 0;
+        boolean isIndexTooLarge = indexOfTask >= this.getTasks().size();
+        if (isNegativeIndex || isIndexTooLarge) {
```
Clear condition statements! Very readable! Good job!

qianz-z commented on src/main/java/io/github/haoyangw/rica/task/TaskManager.java:
```diff
@@ -0,0 +1,240 @@
+        System.out.println(line);
+    }
+
+    private Task rmTask(int indexOfTask) throws RicaTaskException {
```
Maybe you can come up with a better naming for the method?

qianz-z commented on src/main/java/io/github/haoyangw/rica/task/TaskManager.java:
```diff
@@ -0,0 +1,240 @@
+            TaskManager.printlnWithIndent(String.format(TaskManager.TASK_ADDED_PHRASE,
+                    this.getTasks().size()));
+        } else if (howManyTasks == 1) {
+            TaskManager.printlnWithIndent(String.format(TaskManager.SINGLE_TASK_ADDED_PHRASE,
```
I think you may have added an extra line here.

qianz-z commented on src/main/java/io/github/haoyangw/rica/util/StorageManager.java:
```diff
@@ -0,0 +1,92 @@
+        int numOfComponents = pathComponents.length;
+        String fileName = pathComponents[numOfComponents - 1];
+        // beforeLastClash: Basically remove the filename and final '/'
+        int beforeLastSlash = dataPath.length() - fileName.length() - 1;
```
Maybe a better naming would be `indexBeforeLastSlash`?

qianz-z reviewed (commented):
Overall, your iP is well done! However, do take note of magic literals and avoid deep nestings in your code. Good SLAP practices and doing the different levels of abstractions! Keep up the good work!

haoyangw committed:
New TextUi.class: Prints messages Rica-style

Implement some basic helper methods for printing
  the most important messages. If I think of anything
  else, I'll update this class later.

haoyangw committed:
Update classes to print messages with TextUi.class

With TextUi.class, we no longer need to define a
  print message method in every class. Now we just
  initialise TextUi and pass it the message to print.

Contributes to implementing More-OOP increment.

haoyangw committed:
Remove unused variables/methods

We already moved these variables/methods to
  TextUi, so we can remove them from Rica.class
  and TaskManager.class.

haoyangw committed:
Rica.class: Follow proper OOP paradigm

Initialise an instance of Rica and run
  commands from that instance instead of
  directly calling static methods from main()

haoyangw committed:
Rica.class: Make Rica immutable

Mark instance variables final since we
  don't need to mutate Rica during its
  execution lifecycle.

haoyangw committed:
TaskManager.class: Make use of TextUi.printTasks()

We now define the printing syntax for a list
  of Tasks in TextUi, but the actual ArrayList
  of Tasks is in TaskManager.

So, call TaskManager's printTasks() method to
  print tasks as usual, but within
  TaskManager.printTasks(), pass the ArrayList
  of Tasks to TextUi to format the output.

haoyangw committed:
New abstract class: Command.class

Every command supported by Rica will extend
  this superclass. They will all have access
  to the TaskManager(for modifying Tasks) and
  TextUi(to format any output shown to the user).

Run a supported command by initiating the
  corresponding Command and calling run().

haoyangw committed:
Define all of Rica's supported commands as classes

bye, deadline, delete, event, list,
  mark, todo, unmark will now each
  be defined in a subclass of
  Command.class.

As mentioned previously, call .run()
  to execute a specific command.

haoyangw committed:
New CommandManager.class: Manages user commands

Provides helper methods to read the
  next command from the user and parse
  the command into the corresponding
  subclass of Command.class.

If an unrecognised command is issued
  by the user, the new RicaCommandException
  is thrown by CommandManager.parse().

haoyangw committed:
New RicaCommandException.class: Represents command-related exceptions

When a runtime exception occurs due to
  user commands, RicaCommandException
  will be thrown.

haoyangw committed:
Rica.class: Further abstraction

Now that we have CommandManager.class and
  subclasses of Command.class to handle
  user commands, utilise them to read
  and parse commands in Rica.class.

This greatly simplifies the code of Rica.class
  and provides some encapsulation of logic.
  Now, all Rica does is execute the command
  parsed by CommandManager.class, without
  having to distinguish between each command.

haoyangw committed:
Refactor: Rename subpackage 'util' to 'storage'

Much more descriptive this way. We don't have
  any other 'util'-related classes anyway.

haoyangw committed:
TextUi.class: Apply Dependency Inversion Principle

TextUi.printTasks() used to rely on
  ArrayList.class, an implementation of
  the List interface.

Rely on the List interface instead since
  printTasks() only require the generic
  List methods, not anything ArrayList-specific.

haoyangw committed:
TaskManager.class: Add getMatchingTasks() method

Takes in a keyword and returns a List of Tasks
  matching the keyword.

Lays the groundwork for finding a list of Tasks
  matching a keyword for a certain future command.

haoyangw committed:
TextUi.class: Add an overloaded printTasks() method

Support customising the messages shown to the
  user before listing the actual Tasks with an
  overloaded printTasks() method.

This allows printTasks() to be used for more than
  just the 'list' command.

haoyangw committed:
New FindCommand.class: Implements the 'find' command

Supports searching for a list of all
  Tasks matching a keyword given by
  the user.

Example command: 'find book'

haoyangw committed:
CommandManager.class: Support the new 'find' command

Define 'find' as a valid command and
  call the corresponding FindCommand.class
  to execute the command.

haoyangw committed:
Merge pull request #1 from haoyangw/branch-Level-9

Implement increment Level-9

haoyangw committed:
Add JavaDoc comments

haoyangw committed:
Add a missing throws clause and update JavaDoc

Event.deserializeObject() throws RicaSerializationException but
  this is not indicated in the method signature.

Fix this and include an explanation of how RicaSerializationException
  may be thrown.

haoyangw committed:
Remove unnecessary 'throw' clauses

Command.run() handles exceptions on its own, not
  throw them to the caller, so there's no need
  for marking it as 'throws RicaTaskException'.

haoyangw committed:
Merge branch 'master' into branch-A-JavaDoc

haoyangw committed:
Merge pull request #2 from haoyangw/branch-A-JavaDoc

Give Rica proper JavaDoc comments

haoyangw committed:
Add more JavaDoc comments

Include JavaDoc comments for the newly added
  TextUi.printTasks() method.

haoyangw committed:
Improve data file creation logic

Replace my own String parsing code for
  the data file path String with Java's
  own get parent path method Path.getParent()
  to cut down unnecessary coding(duplicate).

Also, use RicaStorageException instead of
  returning an error message String
  from StorageManager.createDataFile().
  More conventional.

Lastly, make the happy path of
  StorageManager.getSavedTasks() clearer
  and include more specific exception
  handling and corresponding user-
  friendly messages to print on the
  user's screen.

haoyangw committed:
Add helper method for printing important Exceptions

TextUi.printImportantErrorMessage() wraps ***
  around an exception's message to make an error
  message look more 'important'.

haoyangw committed:
Add missing throws clause

StorageManager.createDataFile() throws RicaStorageException.
  Indicate as such by marking the method with the 'throws'
  clause.

haoyangw committed:
Add new method for printing important messages

Move the logic for printing an important error
  message to TextUi.printImportantMessage(),
  then call it in TextUi.printImportantErrorMessage()
  using the given RicaException's error message.

haoyangw committed:
Format error message when data file doesn't exist

Use TextUi.printImportantMessage() to make the data file
  missing error message look important.

haoyangw committed:
Format data file doesn't exist message nicely

Print the line separator that all other Rica
  messages use in StorageManager.getSavedTasks()
  as well to standardise Rica's text output.

haoyangw committed:
Add new method to initialiase TaskManager

Move logic for restoring TaskManager's previously state
  from TaskManager's constructor to a new method
  TaskManager.initialise() that Rica can call separately.

This provides 2 benefits:
1) Enables more fine grained control over TaskManager's
  initialisation
2) Provides a layer of abstraction while allowing (1)
     (Rica doesn't care/know what operations 'initialise'
	 involves)

haoyangw committed:
Initialise TaskManager after greeting user

Call TaskManager's new initialise() method only after
  greeting the user so (ugly) error messages(if any)
  are only shown after nicely greeting the user(yay).

haoyangw committed:
Eliminate magic literals

Use named variables in place of all magic literals/numbers
  so code is easier to understand.

haoyangw committed:
Rename methods to be more descriptive

TaskManager.rmTask() (public) is now deleteTaskSpecifiedBy()
  which describes what it does clearly. It also no longer
  conflicts with the private rmTask() method that does a
  different thing(and causes confusion).

We also rename TaskManager.rmTask() (private) to
  rmTaskByIndex() which also describes what it does clearly.

haoyangw committed:
Set the Slate theme for Github Pages

haoyangw committed:
Add ASCII art to Rica's user greeting

Include an ASCII art of R.I.C.A. to Rica's user
  greeting for style.

haoyangw committed:
Fix typo in TaskManager.class

deleteTaskSpecifiedBy() was misspelled
  deleteTaskSpecifiedByl()

haoyangw committed:
Write a detailed User Guide for Rica

haoyangw committed:
Add descriptive error message on Exception thrown

Inform the user something bad happened when
  StorageManager.saveTasks() fails(and apologise of course).

haoyangw committed:
Clear data file if it's corrupted

When Rica finds invalid data in its data file
  while trying to restore her previous memories,
  clear the data file so she doesn't throw another
  exception the next time she is awakened.

haoyangw committed:
Document another exception being thrown

Task.deserializeObject() also throws
  RicaSerializationException, which we ommitted
  from the method's JavaDoc comments and method
  signature. Fix this mistake.

haoyangw committed:
Catch another Exception we missed

StorageManager.getSavedTasks() may also throw
  RicaSerializationExcetion, but we didn't catch this
  in TaskManager.initialise()'s try/catch block, which
  can cause Rica to CRASH.

Prevent this by adding RicaSerializationException
  to the catch block of RicaStorageException
  since they are handled similarly.

haoyangw committed:
Handle NumberFormatException

mark and unmark commands are supposed to be
  passed integer arguments. Currently, if
  non-integers are given, NumberFormatException
  is thrown but not caught.

Handle NumberFormatException by throwing our
  very own RicaCommandException with a
  user-friendly error message that Rica will
  catch(and tell the user about).

haoyangw committed:
Don't capitalise T of 'task'

MarkCommand/UnmarkCommand shouldn't capitalise
  T of 'task' to standardise with TaskManager's
  user messages.

haoyangw committed:
Update the repo README.md
